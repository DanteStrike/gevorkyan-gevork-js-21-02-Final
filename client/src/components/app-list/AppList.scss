@import "src/assets/variables";

.app-list {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  min-height: 300px;

  &--table {
    min-height: 672px;
  }

  &--row-small {
    min-height: unset;
  }

  &--row-small &__nav {
    display: flex;
    justify-content: center;
  }
}

@media (max-width: $tabletWidth) {
  .app-list__nav {
    display: flex;
    justify-content: center;
  }
}
