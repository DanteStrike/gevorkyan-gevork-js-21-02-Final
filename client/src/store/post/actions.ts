import {itemActions} from './slices/item';
import {fetchActions} from './slices/fetch';
import {listActions} from './slices/list';

const actions = {
  fetchActions,
  ...itemActions,
  ...listActions,
};

export default actions;
