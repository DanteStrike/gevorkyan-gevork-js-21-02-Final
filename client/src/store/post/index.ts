export {default as postActions} from './actions';
export {default as postOperations} from './operations';
export {default as postSelectors} from './selectors';
export {default as postReducer} from './reducers';
