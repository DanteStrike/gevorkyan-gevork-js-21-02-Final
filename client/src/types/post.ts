import {IUserPreview} from './user';

export interface IPostPreview {
  id: string;
  text: string;
  image: string;
  publishDate: string;
  owner: IUserPreview;
}
