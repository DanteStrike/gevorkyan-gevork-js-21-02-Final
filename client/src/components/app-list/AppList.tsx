import React from 'react';
import {List} from 'antd';
import './AppList.scss';

export const enum AppListMod {
  TABLE = `table`,
  ROW = `row`,
  SMALL_ROW = `smallRow`,
}

const listLayout = {
  table: {
    gutter: 15,
    xs: 1,
    sm: 2,
    md: 2,
    lg: 3,
    xl: 3,
    xxl: 3,
  },
  row: {
    gutter: 15,
    xs: 1,
    sm: 1,
    md: 3,
    lg: 3,
    xl: 3,
    xxl: 3,
  },
  smallRow: {
    gutter: 15,
    xs: 1,
    sm: 1,
    md: 2,
    lg: 2,
    xl: 2,
    xxl: 2,
  },
};

export interface IAppListProps<T> {
  dataSource: T[];
  renderItem: (item: T, index: number) => React.ReactNode;
  current: number;
  total: number;
  pageSize: number;
  onChange: (page: number) => void;
  isLoading: boolean;
  mod?: AppListMod;
}

function AppList<T>({
  dataSource,
  current,
  total,
  onChange,
  isLoading,
  renderItem,
  pageSize,
  mod = AppListMod.TABLE,
}: IAppListProps<T>) {
  return (
    <List
      className={`app-list ${mod === AppListMod.TABLE ? `app-list--table` : ``} ${
        mod === AppListMod.SMALL_ROW ? `app-list--row-small` : ``
      }`.trim()}
      dataSource={dataSource}
      renderItem={renderItem}
      grid={listLayout[mod]}
      loading={isLoading}
      pagination={{
        defaultCurrent: 1,
        current,
        total,
        showSizeChanger: false,
        onChange,
        disabled: isLoading,
        pageSize,
        className: `app-list__nav`,
      }}
    />
  );
}

AppList.Item = List.Item;

AppList.defaultProps = {
  mod: AppListMod.TABLE,
};

export default AppList;
