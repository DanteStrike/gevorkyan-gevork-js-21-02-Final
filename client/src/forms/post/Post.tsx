import React, {useEffect} from 'react';
import {useHistory} from 'react-router-dom';
import Modal from '../../components/modal/Modal';
import PostCard from '../../components/post-card/PostCard';
import useAppSelector from '../../hooks/use-app-selector';
import {postActions, postOperations, postSelectors} from '../../store/post';
import useAppDispatch from '../../hooks/use-app-dispatch';
import {RequestType} from '../../store/post/slices/fetch';
import {IComment} from '../../types';
import AppList, {AppListMod} from '../../components/app-list/AppList';
import Comment from '../../components/comment/Comment';

interface IPostProps {
  id: string;
}

function Post({id}: IPostProps) {
  const commentsPerPage = 2;

  const dispatch = useAppDispatch();
  const post = useAppSelector(postSelectors.getPost);
  const isPostLoading = useAppSelector(postSelectors.getPostIsLoading);
  const isPostError = useAppSelector(postSelectors.getPostIsError);
  const postError = useAppSelector(postSelectors.getPostError);

  const page = useAppSelector(postSelectors.getPage);
  const total = useAppSelector(postSelectors.getTotal);
  const comments = useAppSelector(postSelectors.getComments);
  const isCommentsLoading = useAppSelector(postSelectors.getCommentsIsLoading);
  const isCommentsError = useAppSelector(postSelectors.getCommentsIsError);
  const commentsError = useAppSelector(postSelectors.getCommentsError);

  const handlePaginationChange = (newPage: number) => {
    dispatch(postActions.changePage(newPage));
  };

  const history = useHistory();
  const handleModalClose = () => {
    const {pathname} = history.location;
    const newRout = pathname.substring(0, pathname.lastIndexOf('/'));
    history.push(newRout);
  };

  useEffect(() => {
    dispatch(postOperations.loadPost(id));
  }, [dispatch, id]);

  useEffect(() => {
    dispatch(postOperations.loadComments(id, commentsPerPage, page));
  }, [dispatch, id, page]);

  useEffect(
    () => () => {
      dispatch(postActions.fetchActions[RequestType.LOAD_POST].requestAbort());
      dispatch(postActions.fetchActions[RequestType.LOAD_COMMENTS].requestAbort());
      dispatch(postActions.reset());
      dispatch(postActions.resetList());
    },
    [dispatch]
  );

  const renderComments = () => (
    <AppList
      current={page}
      total={total}
      dataSource={comments}
      onChange={handlePaginationChange}
      isLoading={isCommentsLoading}
      pageSize={commentsPerPage}
      renderItem={(comment: IComment) => (
        <AppList.Item key={comment.id}>
          <Comment comment={comment} />
        </AppList.Item>
      )}
      mod={AppListMod.SMALL_ROW}
    />
  );

  return (
    <Modal
      isOpen
      onClose={handleModalClose}
      isError={isPostError || isCommentsError}
      errMsg={postError || commentsError}
    >
      <PostCard post={post} modal renderComments={renderComments} isLoading={isPostLoading} />
    </Modal>
  );
}

export default Post;
