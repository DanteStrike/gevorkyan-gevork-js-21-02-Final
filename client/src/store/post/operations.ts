import {AxiosInstance} from 'axios';
import actions from './actions';
import i18next from '../../locale/i18next';
import {IComments, IPostPreview} from '../../types';
import {FetchErrorType} from '../../enums';
import {RequestType} from './slices/fetch';

const loadPost = (id: string) => (dispatch: any, getState: any, api: AxiosInstance) => {
  const controller = new AbortController();

  api
    .get<IPostPreview>(`/post/${id}?locale=${i18next.resolvedLanguage}`, {signal: controller.signal})
    .then((response) => {
      dispatch(actions.set(response.data));
      dispatch(actions.fetchActions[RequestType.LOAD_POST].requestFinished());
    })
    .catch((err) => {
      if (err.response?.data?.error === FetchErrorType.PARAMS_NOT_VALID) {
        dispatch(actions.fetchActions[RequestType.LOAD_POST].requestFailed(i18next.t(`error.network.notFoundPost`)));
      } else {
        dispatch(actions.fetchActions[RequestType.LOAD_POST].requestFailed(err.message));
      }
    });
  dispatch(actions.fetchActions[RequestType.LOAD_POST].requestStart(controller));
};

const loadComments =
  (id: string, limit: number, page: number) => (dispatch: any, getState: any, api: AxiosInstance) => {
    const controller = new AbortController();

    api
      .get<IComments>(`/post/${id}/comment?page=${page - 1}&limit=${limit}&locale=${i18next.resolvedLanguage}`, {
        signal: controller.signal,
      })
      .then((response) => {
        dispatch(actions.setup(response.data));
        dispatch(actions.fetchActions[RequestType.LOAD_COMMENTS].requestFinished());
      })
      .catch((err) => {
        if (err.response?.data?.error === FetchErrorType.PARAMS_NOT_VALID) {
          dispatch(
            actions.fetchActions[RequestType.LOAD_COMMENTS].requestFailed(i18next.t(`error.network.notFoundPost`))
          );
        } else {
          dispatch(actions.fetchActions[RequestType.LOAD_COMMENTS].requestFailed(err.message));
        }
      });
    dispatch(actions.fetchActions[RequestType.LOAD_COMMENTS].requestStart(controller));
  };

export default {
  loadPost,
  loadComments,
};
