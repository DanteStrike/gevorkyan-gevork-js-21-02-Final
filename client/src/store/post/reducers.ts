import {combineReducers} from 'redux';
import {fetchReducer, RequestType} from './slices/fetch';
import {itemReducer} from './slices/item';
import NameSpace from '../name-space';
import {IFetchStore, IListStore} from '../hor';
import {IComment, IPostPreview} from '../../types';
import {listReducer} from './slices/list';

const post = itemReducer;
const reducer = combineReducers({
  fetch: fetchReducer,
  post,
  comments: listReducer,
});
export interface IPostDuckStore {
  [NameSpace.POST]: {
    fetch: {
      [RequestType.LOAD_POST]: IFetchStore;
      [RequestType.LOAD_COMMENTS]: IFetchStore;
    };
    post: IPostPreview;
    comments: IListStore<IComment>;
  };
}
export default reducer;
