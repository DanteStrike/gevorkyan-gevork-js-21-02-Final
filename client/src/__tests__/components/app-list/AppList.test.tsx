import {shallow} from 'enzyme';
import React from 'react';
import AppList from '../../../components/app-list/AppList';
import {posts} from '../../mocks/post.mock';

describe(`Component AppList should render correctly`, () => {
  it(`snapshot default test`, () => {
    const wrapper = shallow(
      <AppList
        onChange={jest.fn()}
        current={1}
        dataSource={posts.data}
        isLoading={false}
        renderItem={jest.fn()}
        pageSize={posts.limit}
        total={posts.total}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
