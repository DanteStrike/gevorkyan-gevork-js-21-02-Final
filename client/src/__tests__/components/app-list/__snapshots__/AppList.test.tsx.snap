// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component AppList should render correctly snapshot default test 1`] = `
<List
  className="app-list app-list--table"
  dataSource={
    Array [
      Object {
        "id": "postID_1",
        "image": "postIMG_URL",
        "owner": Object {
          "firstName": "anyFN",
          "id": "userID_1",
          "lastName": "antLN",
          "picture": "url",
          "title": "ms",
        },
        "publishDate": "un-normalized",
        "text": "postText",
      },
      Object {
        "id": "postID_2",
        "image": "postIMG_URL",
        "owner": Object {
          "firstName": "anyFN",
          "id": "userID_1",
          "lastName": "antLN",
          "picture": "url",
          "title": "ms",
        },
        "publishDate": "un-normalized",
        "text": "postText",
      },
      Object {
        "id": "postID_3",
        "image": "postIMG_URL",
        "owner": Object {
          "firstName": "anyFN",
          "id": "userID_2",
          "lastName": "antLN",
          "picture": "url",
          "title": "ms",
        },
        "publishDate": "un-normalized",
        "text": "postText",
      },
    ]
  }
  grid={
    Object {
      "gutter": 15,
      "lg": 3,
      "md": 2,
      "sm": 2,
      "xl": 3,
      "xs": 1,
      "xxl": 3,
    }
  }
  loading={false}
  pagination={
    Object {
      "className": "app-list__nav",
      "current": 1,
      "defaultCurrent": 1,
      "disabled": false,
      "onChange": [MockFunction],
      "pageSize": 5,
      "showSizeChanger": false,
      "total": 3,
    }
  }
  renderItem={[MockFunction]}
/>
`;
